### PR TITLE
Add support for V2 of the waveshare 5.83in e-paper display.

### LINKED ISSUE #2121: WaveShare e-paper 5.83in V2 display not working
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Hass.io

**ESP (ESP32/ESP8266, Board/Sonoff):**

ESP32 Waveshare driver board. (I also tried with ESP8266 Waveshare driver board resulting in same issue.)

**ESPHome version (latest production, beta, dev branch)**

v1.18

**Affected component:**

WaveShare e-paper display
https://esphome.io/components/display/waveshare_epaper.html

**Description of problem:**
WaveShare 5.83in black & white display does not work with ESPHome, with the display not refreshing at all, and remaining blank (white), with only a black edge appearing around entire screen when connecting to the ESP32 device.  The display has a "V2" sticker on it; could it be that a new model config is needed for that too, as the new 5.83in V2 is now 648x480 resolution rather than 600x448 like the original 5.83in model was?

**Problem-relevant YAML-configuration entries:**
```yaml
esphome:
  name: c-epaper
  platform: ESP32
  board: esp32dev
  on_boot:
    priority: -10
    then:
      - delay: 10s
      - component.update: epaperdisplay

wifi:
  ssid: "removed"
  password: "removed"

# Enable logging
logger:

# Enable Home Assistant API
api:

ota:
  password: "removed"

font:

  - file: 'fonts/Roboto-Regular.ttf'
    id: main_sensor_unit
    size: 20
    glyphs:
      ['&', '@', '!', ',', '.', '"', '%', '(', ')', '+', '-', '_', ':', '°', '0',
       '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E',
       'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
       'T', 'U', 'V', 'W', 'X', 'Y', 'Z', ' ', 'a', 'b', 'c', 'd', 'e', 'f',
       'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
       'u', 'v', 'w', 'x', 'y', 'z','å', 'ä', 'ö', '/','º','µ','³']

spi:
  clk_pin: 13   #SCK
  mosi_pin: 14  #DIN

display:
  - platform: waveshare_epaper
    id: epaperdisplay
    cs_pin: 15
    dc_pin: 27
    update_interval: 600s
    busy_pin: 25
    reset_pin: 26
    model: 5.83in
    lambda: |-

      it.print(40, 40, id(main_sensor_unit), "TEST");

```

**Logs (if applicable):**

```
No relevant log messages / errors.

```

**Additional information and things you've tried:**
I have also tried the display on ESPHome using a Waveshare ESP8266 driver board, but with no better success.

### PR DESCRIPTION
**Pull request by cooki35**
# What does this implement/fix?

Add code to support the WaveShare 5.83in black & white displays with "V2" stickers on them.
The resolution of the V2 version changed (from 600x448) to 648x480.
The init and display code differ from the original 5.83in version.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes [esphome/issues#2121](https://github.com/esphome/issues/issues/2121)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2200

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

The display was connected with the WaveShare e-Paper Driver HAT Rev2.2 and via SPI to an ESP32 (esp32doit-devkit-v1).
The switches of the driver HAT were set to:
- Display Config: B (0.47R)
- Interface Config 0 (4-line SPI)

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
spi:
  clk_pin: GPIO18
  mosi_pin: GPIO23
  id: spi001

display:
  - platform: waveshare_epaper_test
    spi_id: spi001
    cs_pin: GPIO5
    dc_pin: GPIO19
    busy_pin: GPIO16
    reset_pin: GPIO17
    model: 5.83inv2
    update_interval: 30s
    reset_duration: 2ms
    id: eink
    lambda: |-
      it.line(10, 10, 630, 10, COLOR_ON);
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp [hunk 1]
```diff
@@ -937,6 +937,98 @@ void WaveshareEPaper5P8In::dump_config() {
...
+void WaveshareEPaper5P8InV2::initialize() {
+  /*
+    // Reset
+    this->reset_pin_->digital_write(false);
+    delay(20);
+    this->reset_pin_->digital_write(true);
+    delay(5);
+    this->reset_pin_->digital_write(false);
+    delay(20);
+  */
```
Comment (ssieb) [inline esphome/components/waveshare_epaper/waveshare_epaper.cpp]: Why is this commented out?  If it's not needed, then just remove it.

**Review by ssieb — CHANGES_REQUESTED** [review]

**Comment by TheStork239** [comment]
Is work still being done on this?  It would be great to get it over the line, so that the newer 5.83in displays would be usable.  
@cooki35 is the change that @ssieb requested doable?

**Comment by TheStork239** [comment]
This looks like it would be ready to go, apart from the one change requested by the review.
@cooki35, are you planning on making that edit to move this PR forward? It would be great to have the 5.83 v2 supported.

**Comment by izacq** [comment]
Just tested it, works great, thanks @cooki35

**Comment by cooki35** [comment]
The changes sound reasonable, I forgot to delete the comments. I removed them. Thanks @ssieb !

**Review by jesserockz — APPROVED** [review]

**Pull request merged**